core/src/Streamly/Internal/Data/Stream/TH.hs

```diff
@@ -0,0 +1,542 @@
+-- import Streamly.Data.Stream (Stream)
+--
+-- import qualified Streamly.Data.Stream
+-- import qualified Streamly.Data.Stream.Concurrent
+-- import qualified Streamly.Data.Stream.Concurrent as Concur
+-- import qualified Streamly.Internal.Data.Stream.TH as StreamTH
+--
+-- applyConcurrent :: Monad m => Stream m (a -> b) -> Stream m a -> Stream m b
+-- applyConcurrent = Concur.zipWith ($)
+--
+-- \$(StreamTH.makeStreamZipApplicative \"ZipConcurrent\" "applyConcurrent" True)
+-- @
+--
+makeStreamZipApplicative
+    :: String -- ^ Name of the type
+    -> String -- ^ Function to use for (\<*\>)
+    -> Bool   -- ^ 'True' if (\<*\>) has concurrent properties
+    -> Q [Dec]
+makeStreamZipApplicative dtNameStr apOpStr isConcurrent = do
+    typeDec <- makeTypeDec dtNameStr
+    functorI <- makeStreamFunctor dtNameStr
+    applicativeI <-
+        mkStreamApplicative
+            dtNameStr
+            classConstraints
+            "Streamly.Data.Stream.repeat"
+            apOpStr
+    simpleInstances <- mkSimpleStreamInstances dtNameStr
+    return $ typeDec ++ [functorI, applicativeI] ++ simpleInstances
+
+    where
+
+    classConstraints =
+        if isConcurrent
+        then ["Streamly.Internal.Data.Stream.Concurrent.MonadAsync"]
+        else ["Monad"]
+
+-- | Create a type with specific stream combination properties.
+--
+-- __Create a type with serial interleave properties:__
+--
+-- @
+-- import Control.Monad.Catch (MonadThrow(..))
+-- import Control.Monad.IO.Class (MonadIO(..))
+-- import Control.Monad.Reader.Class (MonadReader(..))
+-- import Control.Monad.Trans.Class (MonadTrans(..))
+-- import Streamly.Data.Stream (Stream)
+--
+-- import qualified Control.Monad
+-- import qualified Streamly.Data.Stream
+-- import qualified Streamly.Data.Stream as Stream
+-- import qualified Streamly.Internal.Data.Stream.TH as StreamTH
+--
+-- bindInterleave :: Stream m a -> (a -> Stream m b) -> Stream m b
+-- bindInterleave = flip (Stream.concatMapWith Stream.interleave)
+--
+-- \$(StreamTH.makeStreamMonad \"InterleaveSerial\" "bindInterleave" False)
+-- @
+--
+-- __Create a type with eager concurrency properties:__
+--
+-- @
+-- import Control.Monad.Catch (MonadThrow(..))
+-- import Control.Monad.IO.Class (MonadIO(..))
+-- import Control.Monad.Reader.Class (MonadReader(..))
+-- import Control.Monad.Trans.Class (MonadTrans(..))
+-- import Streamly.Data.Stream (Stream)
+-- import Streamly.Data.Stream.Concurrent (MonadAsync)
+--
+-- import qualified Control.Monad
+-- import qualified Streamly.Data.Stream
+-- import qualified Streamly.Data.Stream.Concurrent
+-- import qualified Streamly.Data.Stream.Concurrent as Concur
+-- import qualified Streamly.Internal.Data.Stream.TH as StreamTH
+--
+-- bindConcurrent :: MonadAsync m => Stream m a -> (a -> Stream m b) -> Stream m b
+-- bindConcurrent = flip (Concur.concatMapWith (Concur.eager True))
+--
+-- \$(StreamTH.makeStreamMonad \"Parallel\" "bindConcurrent" True)
+-- @
+--
+makeStreamMonad
```

Comment:
Can we make it simpler  like `mkMonadType`? Stream is implicit.

Reply:
`Type` is implicit?
`mkMonad`? Sounds weird though.

Reply:
NVM. `TypeGen.mkMonad` sounds good.